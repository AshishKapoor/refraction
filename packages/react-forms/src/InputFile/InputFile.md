The `InputFile` component shares the same API of the HTML `input[type=file]` element.

```js
<InputFile multiple />
```

It also supports all the `InputText` and `Button` properties:

```js
<InputFile size="small" importance="primary" />
```
