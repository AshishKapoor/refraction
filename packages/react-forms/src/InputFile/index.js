// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import callAll from '../utils/callAll';
import mergeRefs from '../utils/mergeRefs';
import InputGroup from '../InputGroup';
import InputText from '../InputText';
import Button from '../Button';

const FileSelector = styled.input`
  opacity: 0;
  position: absolute;
  pointer-events: none;
  width: 0;
  height: 0;
  overflow: hidden;
`;

type Props = {
  onClick?: (SyntheticInputEvent<HTMLInputElement>) => void,
  onChange?: (SyntheticInputEvent<HTMLInputElement>) => void,
  className?: string,
  style?: Object,
  multiple?: boolean,
  innerRef?: React.ElementRef<any>,
};

type State = {
  label: string,
};

class InnerInput extends React.Component<Props, State> {
  state = {
    label: '',
  };

  input = React.createRef();

  handleOnClick = callAll(
    (...args) => this.input.current && this.input.current.click(...args),
    this.props.onClick
  );

  updateLabel = ({ target: { files } }) => {
    const label =
      files && files.length > 0
        ? files.length === 1
          ? files[0].name
          : `${files.length} files selected`
        : '';
    this.setState({ label });
  };

  render() {
    const {
      onClick,
      className,
      style,
      multiple,
      onChange,
      innerRef,
      ...props
    } = this.props;

    return (
      <InputGroup className={className} style={style}>
        {cn => (
          <InputText
            {...props}
            readOnly
            className={cn}
            onClick={this.handleOnClick}
            value={this.state.label}
          />
        )}
        {() => (
          <FileSelector
            type="file"
            ref={mergeRefs(this.input, innerRef)}
            onChange={callAll(this.updateLabel, onChange)}
            multiple={multiple}
          />
        )}
        {cn => (
          <Button
            importance="secondary"
            {...props}
            className={cn}
            onClick={this.handleOnClick}
          >
            Select file
          </Button>
        )}
      </InputGroup>
    );
  }
}

const InputFile: React.StatelessFunctionalComponent<Props> = styled(
  React.forwardRef((props, ref) => <InnerInput {...props} innerRef={ref} />)
)();

// @component
export default InputFile;
