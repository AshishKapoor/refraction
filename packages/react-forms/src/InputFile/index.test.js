/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React from 'react';
import { mount } from 'enzyme';
import InputFile from '.';

jest.mock('../Button', () => 'button');
jest.mock('../InputText', () => 'input');

it('renders an InputFile', () => {
  expect(mount(<InputFile />)).toMatchSnapshot();
});

it('changes label when file is selected', () => {
  const wrapper = mount(<InputFile />);
  expect(
    wrapper
      .find('input')
      .at(0)
      .prop('value')
  ).toBe('');

  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: {
        files: [{ name: 'foobar' }],
      },
    });

  expect(
    wrapper
      .find('input')
      .at(0)
      .prop('value')
  ).toBe('foobar');
});

it('handles cases when the input has no files selected', () => {
  const wrapper = mount(<InputFile />);

  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: {
        files: [],
      },
    });

  expect(
    wrapper
      .find('input')
      .at(0)
      .prop('value')
  ).toBe('');
});

it('triggers onChange event on select', () => {
  const handleChange = jest.fn();
  const wrapper = mount(<InputFile onChange={handleChange} />);

  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: {
        files: [{ name: 'foobar' }],
      },
    });

  expect(handleChange).toHaveBeenCalled();
});

it('generates correct label on multi selection', () => {
  const wrapper = mount(<InputFile />);

  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: {
        files: [{ name: 'foobar' }, { name: 'foobaz' }],
      },
    });

  expect(
    wrapper
      .find('input')
      .at(0)
      .prop('value')
  ).toBe('2 files selected');
});

it('triggers the input file selector on click', () => {
  const handleClick = jest.fn();
  const wrapper = mount(<InputFile onClick={handleClick} />);

  wrapper.find('button').simulate('click');
  expect(handleClick).toHaveBeenCalled();
});
