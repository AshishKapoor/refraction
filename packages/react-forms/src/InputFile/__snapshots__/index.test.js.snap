// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an InputFile 1`] = `
.emotion-5 {
  all: unset;
  display: inline-block;
  border-radius: 2px;
  -webkit-transition: 0.2s ease-in-out border-color;
  transition: 0.2s ease-in-out border-color;
  border: 1px solid #8F9BA3;
}

.emotion-5:focus-within {
  border-color: #00c1bb;
}

.emotion-5[data-invalid='true'] {
  border-color: #E61E27;
}

.emotion-1 {
  opacity: 0;
  position: absolute;
  pointer-events: none;
  width: 0;
  height: 0;
  overflow: hidden;
}

.emotion-0 {
  border-radius: 1px 0 0 1px;
  border-left: 0;
  border-top: 0;
  border-bottom: 0;
}

.emotion-3 {
  border-radius: 0 1px 1px 0;
  border-right: 0;
  border-top: 0;
  border-bottom: 0;
}

<InputFile>
  <ForwardRef
    className="emotion-7 emotion-4"
  >
    <InputGroup
      className="emotion-7 emotion-4"
    >
      <fieldset
        className="emotion-4 emotion-5 emotion-6"
        data-invalid={false}
        onChange={[Function]}
        onInvalid={[Function]}
      >
        <input
          className="emotion-0"
          key=".0"
          onClick={[Function]}
          readOnly={true}
          value=""
        />
        <FileSelector
          key=".1"
          onChange={[Function]}
          type="file"
        >
          <input
            className="emotion-1 emotion-2"
            onChange={[Function]}
            type="file"
          />
        </FileSelector>
        <button
          className="emotion-3"
          importance="secondary"
          key=".2"
          onClick={[Function]}
        >
          Select file
        </button>
      </fieldset>
    </InputGroup>
  </ForwardRef>
</InputFile>
`;
